File: csharp/Wjybxx.Commons.Core/src/Mutable/MutableFloat.cs
#region LICENSE

// Copyright 2024 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;
using System.Collections.Generic;

#pragma warning disable CS1591
namespace Wjybxx.Commons.Mutable;

/// <summary>
/// 可变Int32
/// </summary>
public class MutableFloat : IMutableNumber<float>, IEquatable<MutableFloat>, IComparable<MutableFloat>
{
    private float _value;

    public MutableFloat(float value = 0f) {
        _value = value;
    }

    public float Value {
        get => _value;
        set => _value = value;
    }

    public int IntValue => (int)_value;
    public long LongValue => (long)_value;
    public float FloatValue => _value;
    public double DoubleValue => _value;

    #region op

    /** 加上操作数 */
    public void Add(float operand) {
        this._value += operand;
    }

    /** 返回加上操作数后的值 */
    public float AddAndGet(float operand) {
        this._value += operand;
        return _value;
    }

    /** 返回当前值 */
    public float GetAndAdd(float operand) {
        float last = _value;
        this._value += operand;
        return last;
    }

    /** 加1 */
    public void Increment() {
        _value++;
    }

    /** 返回加1后的值 */
    public float IncrementAndGet() {
        return ++_value;
    }

    /** 加1并返回当前值 */
    public float GetAndIncrement() {
        return _value++;
    }

    /** 减1 */
    public void Decrement() {
        _value--;
    }

    /** 返回减1后的值 */
    public float DecrementAndGet() {
        return --_value;
    }

    /** 减1并返回当前值 */
    public float GetAndDecrement() {
        return _value--;
    }

    #endregion

    #region compare

    public int CompareTo(MutableFloat? other) {
        if (ReferenceEquals(this, other)) return 0;
        if (ReferenceEquals(null, other)) return 1;
        return _value.CompareTo(other._value);
    }

    public static bool operator <(MutableFloat? left, MutableFloat? right) {
        return Comparer<MutableFloat>.Default.Compare(left, right) < 0;
    }

    public static bool operator >(MutableFloat? left, MutableFloat? right) {
        return Comparer<MutableFloat>.Default.Compare(left, right) > 0;
    }

    public static bool operator <=(MutableFloat? left, MutableFloat? right) {
        return Comparer<MutableFloat>.Default.Compare(left, right) <= 0;
    }

    public static bool operator >=(MutableFloat? left, MutableFloat? right) {
        return Comparer<MutableFloat>.Default.Compare(left, right) >= 0;
    }

    #endregion

    #region equals

    public bool Equals(MutableFloat? other) {
        if (other == null) {
            return false;
        }
        return _value.Equals(other._value);
    }

    public override bool Equals(object? obj) {
        if (ReferenceEquals(null, obj)) return false;
        if (ReferenceEquals(this, obj)) return true;
        if (obj.GetType() != this.GetType()) return false;
        return Equals((MutableFloat)obj);
    }

    public override int GetHashCode() {
        return _value.GetHashCode();
    }

    public static bool operator ==(MutableFloat? left, MutableFloat? right) {
        return Equals(left, right);
    }

    public static bool operator !=(MutableFloat? left, MutableFloat? right) {
        return !Equals(left, right);
    }

    public override string ToString() {
        return $"{nameof(Value)}: {Value}";
    }

    #endregion
}

File: java/Commons-Base/src/main/java/cn/wjybxx/base/mutable/MutableFloat.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.base.mutable;

/**
 * @author wjybxx
 * date - 2024/1/4
 */
public class MutableFloat extends Number implements MutableNumber<Float>, Comparable<MutableFloat> {

    private float value;

    public MutableFloat() {
    }

    public MutableFloat(float value) {
        this.value = value;
    }

    public MutableFloat(Number value) {
        this.value = value.floatValue();
    }

    public MutableFloat(String value) {
        this.value = Float.parseFloat(value);
    }

    // region cast

    @Override
    public Float getValue() {
        return value;
    }

    @Override
    public void setValue(Float value) {
        this.value = value;
    }

    public void setValue(final float value) {
        this.value = value;
    }

    @Override
    public int intValue() {
        return (int) value;
    }

    @Override
    public long longValue() {
        return (long) value;
    }

    @Override
    public float floatValue() {
        return value;
    }

    @Override
    public double doubleValue() {
        return value;
    }

    // endregion


    // region op

    /** 加上操作数 */
    public void add(final float operand) {
        this.value += operand;
    }

    /** 返回加上操作数后的值 */
    public float addAndGet(final float operand) {
        this.value += operand;
        return value;
    }

    /** 返回当前值 */
    public float getAndAdd(final float operand) {
        final float last = value;
        this.value += operand;
        return last;
    }

    /** 加1 */
    public void increment() {
        value++;
    }

    /** 返回加1后的值 */
    public float incrementAndGet() {
        return ++value;
    }

    /** 加1并返回当前值 */
    public float getAndIncrement() {
        return value++;
    }

    /** 减1 */
    public void decrement() {
        value--;
    }

    /** 返回减1后的值 */
    public float decrementAndGet() {
        return --value;
    }

    /** 减1并返回当前值 */
    public float getAndDecrement() {
        return value--;
    }

    @Override
    public MutableFloat add(Number operand) {
        this.value += operand.floatValue();
        return this;
    }

    @Override
    public MutableFloat subtract(Number operand) {
        this.value -= operand.floatValue();
        return this;
    }

    @Override
    public MutableFloat multiply(Number operand) {
        this.value *= operand.floatValue();
        return this;
    }

    @Override
    public MutableFloat divide(Number operand) {
        this.value /= operand.floatValue();
        return this;
    }

    // endregion

    // region equals

    @Override
    public int compareTo(MutableFloat that) {
        return Float.compare(value, that.value);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        MutableFloat that = (MutableFloat) o;
        return value == that.value;
    }

    @Override
    public int hashCode() {
        return Float.hashCode(value);
    }

    @Override
    public String toString() {
        return "MutableFloat{" +
                "value=" + value +
                '}';
    }

    // endregion
}
